# 🛒 **연관 분석 (Association Rule Analysis) 과제**

### <span style="color:black; background-color:#F5F5F5;"> **Q1. 연관 규칙 {우유} → {쿠키}가 도출되었을 때, 다음 용어들이 각각 무엇을 의미하는지 설명하시오.**
#### <span style="color:black; background-color:#F5F5F5;">**① 지지도(support) ② 신뢰도(confidence) ③ 향상도(lift)** </span>

답: 
1. 지지도: 전체 거래에서 우유와 쿠키를 같이 샀을 확률
2. 신뢰도: 우유를 구매했을 때, 쿠키도 구매할 확률 (조건부 확률)
3. 우유를 구매했을 때, 쿠키를 구매할 확률 / 그냥 쿠키를 구매할 확률. 즉, 쿠키만 구매한 경우에 대비해 우유를 통해 쿠키로 유입된 경우의 비율

### <span style="color:black; background-color:#F5F5F5;"> **Q2. Apriori 알고리즘이 처리해야 할 후보 항목 수가 기하급수적으로 증가하는 이유와, FP-Growth가 이를 어떻게 해결하는지 설명하시오.**

답: apriori는 최소 지지도를 기준으로 모든 항목들에 대해 필터링 작업을 한다. 그리고 최소 지지도를 넘는 항목들에 대해 추가적으로 필터링을 실시하는데, 이러한 과정 속에서 계속해서 각 항목 간의 지지도 계산을 해야하기 때문에 처리해야 할 후보 항목 수가 기하급수적으로 증가한다.
FP-Growth는 데이터 스캔을 두 번만 하며 이러한 문제를 해결한다. 첫 스캔은 apriori와 같이 최소 지지도를 넘지 않는 항목들을 제거하고 빈도수 기준으로 내림차순 정렬한다. 두 번째 스캔을 통해 FP-tree를 구성한다. 즉, 2번의 스캔을 통해서 알고리즘을 구현하기때문에 계산량이 줄어들게 된다.

# <span style="color:black; background-color:#F5F5F5;"> 💸 **연관 분석을 활용한 잉마트(Ing-Mart) 고객 장바구니 패턴 분석 및 비즈니스 전략 수립** </span>

<strong>죽지도 않고 다시 돌아온 잉마트..! 🤣🫥😫🙃<br>
이번 연관 분석 심화 세션의 과제는 잉마트의 고객 장바구니 분석과 전략 수립입니다~ <strong>



<span style="color:black; background-color:#E6E6FA; padding:2px 4px; border-radius:4px">
<strong> 🤓 지난 심화 세션에서 배운 개념과 실습 내용을 바탕으로 아래 빈칸을 채워주시고, 해당 장바구니 결과를 분석하여 이에 적합한 전략을 제시해주시면 됩니다! <strong>
</span>

## **1️⃣ 데이터 불러오기 및 전처리**

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
from mlxtend.frequent_patterns import fpgrowth
import warnings
warnings.simplefilter(action='ignore', category=DeprecationWarning)

In [3]:
# !pip install mlxtend

- 데이터 불러오기
  - 1차 인사이콘 때 활용하셨던 데이터 원본을 활용해주시면 됩니다!
  - 알맞게 경로 지정해주세요~

In [4]:
transaction_data = pd.read_csv('transaction_data.csv')
product = pd.read_csv('product.csv')

In [5]:
transaction_data.head()

Household_ID    Basket_ID  Product_ID  Store_ID  Day  Quantity  \
0          1803  30780785930     1065887       338  252         1   
1          2299  33768622588     1073244       446  456         1   
2           158  30202616809     7025114       343  225         1   
3          2347  42076926172     1064299       438  695         1   
4          1430  31625201009     1040197     31742  312         1   

   Sales_Value  Trans_time  Week_no  Disc(retail)  Disc(coupon)  \
0         8.99        1419       37         -1.00           0.0   
1         1.00        2030       66         -1.59           0.0   
2         1.50        1246       33         -0.50          -1.0   
3         2.50        1430      100         -2.89           0.0   
4         2.29        1423       45          0.00           0.0   

   Disc(coupon_match)  
0                 0.0  
1                 0.0  
2                 0.0  
3                 0.0  
4                 0.0

- 이상치 처리
  - 예시로 제시해드리지만 추가적으로 필요한 부분은 전처리 처리해주세요!

In [6]:
# Quantity 열에 대해 Z-score를 계산한 뒤, 절댓값을 취해 새로운 열 'z_score'에 저장
transaction_data["z_score"] = np.abs(stats.zscore(transaction_data["Quantity"]))

# Z-score가 3을 초과하는 이상치(즉, 평균에서 3표준편차 이상 벗어난 값)를 추출
outliers_zscore = transaction_data[transaction_data["z_score"] > 3]

- 고객 - 상품 행렬 생성

In [7]:
# 고객-상품 pivot_table 생성 (행: 고객, 열: 상품, 값: 총 구매금액)
user_item_matrix = transaction_data.pivot_table(
    index='Household_ID',     # 가구 ID 기준
    columns='Product_ID',     # 상품 ID 기준
    values='Sales_Value',     # 구매 금액
    aggfunc='sum',            # 상품별 총 구매금액
    fill_value=0              # 구매 이력 없으면 0
)
user_item_matrix

Product_ID    25671     26081     26093     26190     26355     26426     \
Household_ID                                                               
1                  0.0       0.0       0.0       0.0       0.0       0.0   
2                  0.0       0.0       0.0       0.0       0.0       0.0   
3                  0.0       0.0       0.0       0.0       0.0       0.0   
4                  0.0       0.0       0.0       0.0       0.0       0.0   
5                  0.0       0.0       0.0       0.0       0.0       0.0   
...                ...       ...       ...       ...       ...       ...   
2496               0.0       0.0       0.0       0.0       0.0       0.0   
2497               0.0       0.0       0.0       0.0       0.0       0.0   
2498               0.0       0.0       0.0       0.0       0.0       0.0   
2499               0.0       0.0       0.0       0.0       0.0       0.0   
2500               0.0       0.0       0.0       0.0       0.0       0.0   

Product_ID    26540     26601     26636     26691     ...  18273019  18273051  \
Household_ID                                          ...                       
1                  0.0       0.0       0.0       0.0  ...       0.0       0.0   
2                  0.0       0.0       0.0       0.0  ...       0.0       0.0   
3                  0.0       0.0       0.0       0.0  ...       0.0       0.0   
4                  0.0       0.0       0.0       0.0  ...       0.0       0.0   
5                  0.0       0.0       0.0       0.0  ...       0.0       0.0   
...                ...       ...       ...       ...  ...       ...       ...   
2496               0.0       0.0       0.0       0.0  ...       0.0       0.0   
2497               0.0       0.0       0.0       0.0  ...       0.0       0.0   
2498               0.0       0.0       0.0       0.0  ...       0.0       0.0   
2499               0.0       0.0       0.0       0.0  ...       0.0       0.0   
2500               0.0       0.0       0.0       0.0  ...       0.0       0.0   

Product_ID    18273115  18273133  18292005  18293142  18293439  18293696  \
Household_ID                                                               
1                  0.0       0.0       0.0       0.0       0.0       0.0   
2                  0.0       0.0       0.0       0.0       0.0       0.0   
3                  0.0       0.0       0.0       0.0       0.0       0.0   
4                  0.0       0.0       0.0       0.0       0.0       0.0   
5                  0.0       0.0       0.0       0.0       0.0       0.0   
...                ...       ...       ...       ...       ...       ...   
2496               0.0       0.0       0.0       0.0       0.0       0.0   
2497               0.0       0.0       0.0       0.0       0.0       0.0   
2498               0.0       0.0       0.0       0.0       0.0       0.0   
2499               0.0       0.0       0.0       0.0       0.0       0.0   
2500               0.0       0.0       0.0       0.0       0.0       0.0   

Product_ID    18294080  18316298  
Household_ID                      
1                  0.0       0.0  
2                  0.0       0.0  
3                  0.0       0.0  
4                  0.0       0.0  
5                  0.0       0.0  
...                ...       ...  
2496               0.0       0.0  
2497               0.0       0.0  
2498               0.0       0.0  
2499               0.0       0.0  
2500               0.0       0.0  

[2500 rows x 92339 columns]

In [8]:
# 행렬의 크기 확인 (고객 수 × 상품 수)
user_item_matrix.shape

(2500, 92339)

- 구매가 적은 사용자/상품 필터링

In [9]:
# 필터링 기준 정의
min_product_purchases = 10   # 최소 10명 이상이 구매한 상품만 사용
min_user_purchases = 2       # 최소 2개 이상 상품 구매한 사용자만 사용

# 상품별 구매된 고객 수 계산
product_purchase_count = (user_item_matrix > 0).sum()

# 고객별 구매한 상품 수 계산
user_purchase_count = (user_item_matrix > 0).sum(axis=1)

# 기준을 만족하는 상품과 사용자 필터링
filtered_products = product_purchase_count[product_purchase_count >= min_product_purchases].index
filtered_users = user_purchase_count[user_purchase_count >= min_user_purchases].index

# 필터링된 행렬 추출
filtered_matrix = user_item_matrix.loc[filtered_users, filtered_products]
print(f"\n2. Filtered Matrix Shape: {filtered_matrix.shape}")


2. Filtered Matrix Shape: (2500, 23326)


- 이상치 및 음수 제거한 트랜잭션 데이터 생성

In [10]:
# Z-score 기준으로 이상치 제거 (±3 이상) + 구매 수량이 양수인 데이터만 남김
transaction_data_cleaned = transaction_data[
    (transaction_data['z_score'] < 3) & 
    (transaction_data['z_score'] > -3) & 
    (transaction_data['Quantity'] > 0)
]
print(transaction_data_cleaned.shape)

# 데이터 샘플 확인
transaction_data_cleaned.head()

(2559321, 13)


Household_ID    Basket_ID  Product_ID  Store_ID  Day  Quantity  \
0          1803  30780785930     1065887       338  252         1   
1          2299  33768622588     1073244       446  456         1   
2           158  30202616809     7025114       343  225         1   
3          2347  42076926172     1064299       438  695         1   
4          1430  31625201009     1040197     31742  312         1   

   Sales_Value  Trans_time  Week_no  Disc(retail)  Disc(coupon)  \
0         8.99        1419       37         -1.00           0.0   
1         1.00        2030       66         -1.59           0.0   
2         1.50        1246       33         -0.50          -1.0   
3         2.50        1430      100         -2.89           0.0   
4         2.29        1423       45          0.00           0.0   

   Disc(coupon_match)   z_score  
0                 0.0  0.086202  
1                 0.0  0.086202  
2                 0.0  0.086202  
3                 0.0  0.086202  
4                 0.0  0.086202

- 정제된 데이터로 다시 사용자-상품 행렬 생성 및 필터링

In [11]:
# 더 정확한 연관 규칙 도출을 위해 이상치 제거 후 재생성
# 정제된 데이터를 기반으로 사용자-상품 매트릭스 다시 생성
user_item_matrix = transaction_data_cleaned.pivot_table(
    index='Household_ID',     # 가구 ID 기준
    columns='Product_ID',     # 상품 ID 기준
    values='Sales_Value',     # 구매 금액
    aggfunc='sum',            # 상품별 총 구매금액
    fill_value=0              # 구매 이력 없으면 0
)

# 필터링 기준 재사용
min_product_purchases = 10  
min_user_purchases = 2     

# 상품/사용자별 구매 횟수 계산
product_purchase_count = (user_item_matrix > 0).sum()
user_purchase_count = (user_item_matrix > 0).sum(axis=1)

# 조건에 맞는 상품과 사용자 필터
filtered_products = product_purchase_count[product_purchase_count >= min_product_purchases].index
filtered_users = user_purchase_count[user_purchase_count >= min_user_purchases].index

# 최종 필터링된 행렬 생성
filtered_matrix = user_item_matrix.loc[filtered_users, filtered_products]
filtered_matrix

Product_ID    27658     34873     43020     43871     59666     138619    \
Household_ID                                                               
1                  0.0       0.0       0.0       0.0       0.0       0.0   
2                  0.0       0.0       0.0       0.0       0.0       0.0   
3                  0.0       0.0       0.0       0.0       0.0       0.0   
4                  0.0       0.0       0.0       0.0       0.0       0.0   
5                  0.0       0.0       0.0       0.0       0.0       0.0   
...                ...       ...       ...       ...       ...       ...   
2496               0.0       0.0       0.0       0.0       0.0       0.0   
2497               0.0       0.0       0.0       0.0       0.0       0.0   
2498               0.0       0.0       0.0       0.0       0.0       0.0   
2499               0.0       0.0       0.0       0.0       0.0       0.0   
2500               0.0       0.0       0.0       0.0       0.0       0.0   

Product_ID    197681    201704    215923    244960    ...  18005913  18005929  \
Household_ID                                          ...                       
1                  0.0       0.0       0.0      0.00  ...       0.0       0.0   
2                  0.0       0.0       0.0      3.19  ...       0.0       0.0   
3                  0.0       0.0       0.0      0.00  ...       0.0       0.0   
4                  0.0       0.0       0.0      0.00  ...       0.0       0.0   
5                  0.0       0.0       0.0      0.00  ...       0.0       0.0   
...                ...       ...       ...       ...  ...       ...       ...   
2496               0.0       0.0       0.0      0.00  ...       0.0       0.0   
2497               0.0       0.0       0.0      0.00  ...       0.0       0.0   
2498               0.0       0.0       0.0      0.00  ...       0.0       0.0   
2499               0.0       0.0       0.0      0.00  ...       0.0       0.0   
2500               0.0       0.0       0.0      0.00  ...       0.0       0.0   

Product_ID    18022252  18055205  18055329  18105264  18106286  18119016  \
Household_ID                                                               
1                  0.0       0.0       0.0       0.0       0.0       0.0   
2                  0.0       0.0       0.0       0.0       0.0       0.0   
3                  0.0       0.0       0.0       0.0       0.0       0.0   
4                  0.0       0.0       0.0       0.0       0.0       0.0   
5                  0.0       0.0       0.0       0.0       0.0       0.0   
...                ...       ...       ...       ...       ...       ...   
2496               0.0       0.0       0.0       0.0       0.0       0.0   
2497               0.0       0.0       0.0       0.0       0.0       0.0   
2498               0.0       0.0       0.0       0.0       0.0       0.0   
2499               0.0       0.0       0.0       0.0       0.0       0.0   
2500               0.0       0.0       0.0       0.0       0.0       0.0   

Product_ID    18147612  18203921  
Household_ID                      
1                  0.0       0.0  
2                  0.0       0.0  
3                  0.0       0.0  
4                  0.0       0.0  
5                  0.0       0.0  
...                ...       ...  
2496               0.0       0.0  
2497               0.0       0.0  
2498               0.0       0.0  
2499               0.0       0.0  
2500               0.0       0.0  

[2500 rows x 23320 columns]

- 상품 정보 조인하여 제품 타입 단위로 분석 준비

In [12]:
# Product_ID을 기준으로 데이터 product와 inner join하여 Product_type 정보 추가
merged_data = pd.merge(transaction_data_cleaned, product, on='Product_ID', how='inner')
merged_data

Household_ID    Basket_ID  Product_ID  Store_ID  Day  Quantity  \
0                1803  30780785930     1065887       338  252         1   
1                2299  33768622588     1073244       446  456         1   
2                 158  30202616809     7025114       343  225         1   
3                2347  42076926172     1064299       438  695         1   
4                1430  31625201009     1040197     31742  312         1   
...               ...          ...         ...       ...  ...       ...   
2559316           225  33655617762     6463742     31642  447         1   
2559317           266  40097402794      968269       367  551         1   
2559318          2364  30443976140     2041688       673  230         2   
2559319          1048  31145396229     1082185       436  280         1   
2559320          2270  28318130093     1138858       445  111         1   

         Sales_Value  Trans_time  Week_no  Disc(retail)  Disc(coupon)  \
0               8.99        1419       37         -1.00           0.0   
1               1.00        2030       66         -1.59           0.0   
2               1.50        1246       33         -0.50          -1.0   
3               2.50        1430      100         -2.89           0.0   
4               2.29        1423       45          0.00           0.0   
...              ...         ...      ...           ...           ...   
2559316         2.29        1942       65         -1.70           0.0   
2559317         1.99        2255       79          0.00           0.0   
2559318         1.50        1757       34         -0.62           0.0   
2559319         0.74        1445       41          0.00           0.0   
2559320         2.69         808       17         -0.30           0.0   

         Disc(coupon_match)   z_score  Manufacturer     Brand Category  \
0                       0.0  0.086202            69   Private  DRUG GM   
1                       0.0  0.086202          2110  National  GROCERY   
2                       0.0  0.086202          1838  National  GROCERY   
3                       0.0  0.086202          2193  National  GROCERY   
4                       0.0  0.086202           869  National  GROCERY   
...                     ...       ...           ...       ...      ...   
2559316                 0.0  0.086202            69   Private  GROCERY   
2559317                 0.0  0.086202          1091  National  GROCERY   
2559318                 0.0  0.085335            69   Private  GROCERY   
2559319                 0.0  0.086202             2  National  PRODUCE   
2559320                 0.0  0.086202           151  National  GROCERY   

                    Subcategory                    Product_type  \
0                    ANALGESICS                ADULT ANALGESICS   
1        ICE CREAM/MILK/SHERBTS                   PREMIUM PINTS   
2        BAKED BREAD/BUNS/ROLLS  MAINSTREAM WHEAT/MULTIGRAIN BR   
3        ICE CREAM/MILK/SHERBTS                         PREMIUM   
4                          TEAS      TEA BAGS HERBAL & FLAVORED   
...                         ...                             ...   
2559316                  CHEESE                 SHREDDED CHEESE   
2559317  HOUSEHOLD CLEANG NEEDS              TOILET BOWL MANUAL   
2559318     MEAT - SHELF STABLE                   CHILI: CANNED   
2559319          TROPICAL FRUIT                         BANANAS   
2559320            FROZEN PIZZA                   PIZZA/PREMIUM   

        Curr_Size_of_Product  
0                             
1                         PT  
2                      20 OZ  
3                      48 OZ  
4                      20 CT  
...                      ...  
2559316                16 OZ  
2559317                24 OZ  
2559318                15 OZ  
2559319                40 LB  
2559320              12.5 OZ  

[2559321 rows x 19 columns]

- 장바구니(Basket_ID)별로 구매한 상품타입 목록 정리

In [13]:
# 각 거래(Basket_ID)마다 구매한 상품 유형(Product_type)의 리스트 생성
transactions = merged_data.groupby('Basket_ID')['Product_type'].unique().reset_index()
transactions

Basket_ID                                       Product_type
0       26984851472  [POTATOES RUSSET (BULK&BAG), CELERY, ONIONS SW...
1       26984851516  [HAMBURGER BUNS, TRAY PACK/CHOC CHIP COOKIES, ...
2       26984896261  [GRANOLA BARS, EGGS - X-LARGE, LINKS - RAW, SN...
3       26984905972  [RAMEN NOODLES/RAMEN CUPS, MAINSTREAM WHITE BR...
4       26984945254  [SEASONAL CANDY BOX NON-CHOCOLA, INSIDE FROST ...
...             ...                                                ...
254725  42302712006  [TORTILLA/NACHO CHIPS, SFT DRNK 2 LITER BTL CA...
254726  42302712189  [PLSTC CTLRYTBLCLTHSTTHPKSST, REFRIG DIPS, PAP...
254727  42302712298  [BUTTER, HAIR CONDITIONERS AND RINSES, SHAMPOO...
254728  42305362497  [SEASONAL MISCELLANEOUS, SFT DRNK 2 LITER BTL ...
254729  42305362535  [BEERALEMALT LIQUORS, MACARONI & CHEESE DNRS, ...

[254730 rows x 2 columns]

- 트랜잭션 리스트로 변환

In [14]:
transaction_list = transactions['Product_type'].tolist()
transaction_list = [list(item) for item in transaction_list]
print(transaction_list[:5])

[['POTATOES RUSSET (BULK&BAG)', 'CELERY', 'ONIONS SWEET (BULK&BAG)', 'ORGANIC CARROTS', 'BANANAS'], ['HAMBURGER BUNS', 'TRAY PACK/CHOC CHIP COOKIES', 'PEANUT BUTTER', 'SPONGES: BATH HOUSEHOLD', 'GRAHAM CRACKERS'], ['GRANOLA BARS', 'EGGS - X-LARGE', 'LINKS - RAW', 'SNACK CRACKERS', 'GRND/PATTY - ROUND'], ['RAMEN NOODLES/RAMEN CUPS', 'MAINSTREAM WHITE BREAD'], ['SEASONAL CANDY BOX NON-CHOCOLA', 'INSIDE FROST BULBS', 'CHEWING GUM']]


- 트랜잭션 통계

In [15]:
transactions['num_products'] = transactions['Product_type'].apply(len)
average_products_per_order = transactions['num_products'].mean()
max_products_per_order = transactions['num_products'].max()
min_products_per_order = transactions['num_products'].min()

print(f"Average number of products per order: {average_products_per_order}")
print(f"Maximum number of products per order: {max_products_per_order}")
print(f"Minimum number of products per order: {min_products_per_order}")

Average number of products per order: 8.520123267773721
Maximum number of products per order: 128
Minimum number of products per order: 1


<span style="color:black; background-color:#E6E6FA; padding:2px 4px; border-radius:4px">
<strong> 🤓 지금부턴 지난 실습 때 했던 과정의 반복! <strong>
</span>

## **2️⃣ 연관 분석 - TransactionEncoder로 이진 행렬로 변환**

In [16]:
from mlxtend.preprocessing import TransactionEncoder

te = TransactionEncoder()
te_ary = te.fit(transaction_list).transform(transaction_list)  # 학습과 변환을 따로따로!

df_encoded = pd.DataFrame(te_ary, columns=te.columns_).astype(int)
df_encoded.head()

*ATH ACCES:TOWEL BARS/SOAP D  *ATTERIES:CAMERA/FLASH/WATCH  \
0                             0                             0   
1                             0                             0   
2                             0                             0   
3                             0                             0   
4                             0                             0   

   *BOYS/GIRLS MISC TOYS  *MISC. LOBBY ITEMS  *PURSES UMBRELLAS  *SCRAPBOOK  \
0                      0                   0                  0           0   
1                      0                   0                  0           0   
2                      0                   0                  0           0   
3                      0                   0                  0           0   
4                      0                   0                  0           0   

   *SPORT NOVELTIES  *SPORTS APPAREL  *SPRING/SUMMER MISC  20 BIKES  ...  \
0                 0                0                    0         0  ...   
1                 0                0                    0         0  ...   
2                 0                0                    0         0  ...   
3                 0                0                    0         0  ...   
4                 0                0                    0         0  ...   

   WRITING INSTRUMENTS  XMAS PLUSH  YARDLEY  YEAST: DRY  YELLOW JACKET  \
0                    0           0        0           0              0   
1                    0           0        0           0              0   
2                    0           0        0           0              0   
3                    0           0        0           0              0   
4                    0           0        0           0              0   

   YELLOW SUMMER SQUASH  YNG MEN SCREEN PRINT T-SHIRTS  YOGURT  \
0                     0                              0       0   
1                     0                              0       0   
2                     0                              0       0   
3                     0                              0       0   
4                     0                              0       0   

   YOGURT MULTI-PACKS  YOGURT NOT MULTI-PACKS  
0                   0                       0  
1                   0                       0  
2                   0                       0  
3                   0                       0  
4                   0                       0  

[5 rows x 2373 columns]

## **3️⃣ 연관 분석 - Apriori 알고리즘 활용**

- Apriori 알고리즘으로 빈발 항목 집합 도출 (지지도 0.05% 이상)

In [17]:
from mlxtend.frequent_patterns import apriori

filtered_onehot = df_encoded.loc[:, df_encoded.sum(axis=0) > 20]

# apriori로 frequent_itemsets 추출 (최소지지도는 0.005, use_colnames=True, low_memory=True)
frequent_itemsets = apriori(filtered_onehot, min_support=0.005, use_colnames=True, low_memory=True)
frequent_itemsets.head()

support                        itemsets
0  0.011899              (ADULT ANALGESICS)
1  0.030020                  (ADULT CEREAL)
2  0.005374  (AIR CARE - CONTINUOUS ACTION)
3  0.008024            (ALKALINE BATTERIES)
4  0.040266             (ALL FAMILY CEREAL)

- `association_rules`로 연관 규칙 도출 및 필터링(신뢰도 40% 이상)

In [18]:
num_itemsets = len(frequent_itemsets)

# confidence 기준으로 연관 규칙(rules)을 추출하세요.
# 조건: min_threshold=0.4, metric="***",num_itemsets = num_itemsets
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.4, num_itemsets = num_itemsets)

rules[['antecedents', 'consequents', 'support', 'confidence', 'lift']].head()

antecedents              consequents   support  \
0                   (ADULT CEREAL)                (BANANAS)  0.012119   
1                   (ADULT CEREAL)  (FLUID MILK WHITE ONLY)  0.019154   
2              (ALL FAMILY CEREAL)  (FLUID MILK WHITE ONLY)  0.026137   
3  (APPLE JUICE & CIDER (OVER 50%)  (FLUID MILK WHITE ONLY)  0.006874   
4         (APPLES GALA (BULK&BAG))                (BANANAS)  0.006996   

   confidence      lift  
0    0.403688  3.393439  
1    0.638028  2.650870  
2    0.649118  2.696946  
3    0.588769  2.446211  
4    0.507837  4.268928

In [19]:
print(rules.shape)

(445, 14)


- 연관 분석용 리스트 구조 정리

In [20]:
transaction_list = [list(t) if isinstance(t, (list, np.ndarray)) else [t] for t in transaction_list]
transaction_list = [t.tolist() if isinstance(t, np.ndarray) else list(t) if isinstance(t, list) else [t] for t in transaction_list]

- 불필요한 지표 제거

In [21]:
apriori = rules.drop(columns=[
    "antecedent support", 
    "consequent support", 
    "representativity", 
    "conviction", 
    "zhangs_metric", 
    "jaccard", 
    "certainty", 
    "kulczynski"
])

- 유의미한 규칙 필터링(향상도 1 이상)

In [22]:
apriori = apriori[apriori['lift'] >= 1]

- 결과 확인

In [23]:
apriori

antecedents  \
0                                       (ADULT CEREAL)   
1                                       (ADULT CEREAL)   
2                                  (ALL FAMILY CEREAL)   
3                      (APPLE JUICE & CIDER (OVER 50%)   
4                             (APPLES GALA (BULK&BAG))   
..                                                 ...   
440                     (KIDS CEREAL, SHREDDED CHEESE)   
441      (KIDS CEREAL, SOFT DRINKS 12/18&15PK CAN CAR)   
442                    (SHREDDED CHEESE, POTATO CHIPS)   
443  (SOFT DRINKS 12/18&15PK CAN CAR, SNACK CAKE - ...   
444  (BANANAS, SOFT DRINKS 12/18&15PK CAN CAR, MAIN...   

                  consequents   support  confidence      lift  leverage  
0                   (BANANAS)  0.012119    0.403688  3.393439  0.008547  
1     (FLUID MILK WHITE ONLY)  0.019154    0.638028  2.650870  0.011928  
2     (FLUID MILK WHITE ONLY)  0.026137    0.649118  2.696946  0.016446  
3     (FLUID MILK WHITE ONLY)  0.006874    0.588769  2.446211  0.004064  
4                   (BANANAS)  0.006996    0.507837  4.268928  0.005357  
..                        ...       ...         ...       ...       ...  
440  (MAINSTREAM WHITE BREAD)  0.005076    0.442202  4.200561  0.003868  
441  (MAINSTREAM WHITE BREAD)  0.005135    0.454798  4.320211  0.003946  
442  (MAINSTREAM WHITE BREAD)  0.005602    0.405974  3.856424  0.004149  
443  (MAINSTREAM WHITE BREAD)  0.005429    0.478878  4.548949  0.004236  
444   (FLUID MILK WHITE ONLY)  0.005013    0.721877  2.999244  0.003342  

[445 rows x 6 columns]

## **[ 참고 ] 연관 분석 - FP-Growth 알고리즘 활용**

- 코드를 돌릴 때 조심해주세요!

In [24]:
filtered_onehot

ABRASIVES  ACNE MEDICATIONS  ACTIVITY  ADDITIVES/FLUIDS  \
0               0                 0         0                 0   
1               0                 0         0                 0   
2               0                 0         0                 0   
3               0                 0         0                 0   
4               0                 0         0                 0   
...           ...               ...       ...               ...   
254725          0                 0         0                 0   
254726          0                 0         0                 0   
254727          0                 0         0                 0   
254728          0                 0         0                 0   
254729          0                 0         0                 0   

        ADHESIVES/CAULK  ADULT ANALGESICS  ADULT CEREAL  \
0                     0                 0             0   
1                     0                 0             0   
2                     0                 0             0   
3                     0                 0             0   
4                     0                 0             0   
...                 ...               ...           ...   
254725                0                 0             0   
254726                0                 0             0   
254727                0                 0             0   
254728                0                 0             0   
254729                0                 0             0   

        ADULT INCONTINENCE BRIEFS  ADULT INCONTINENCE MISC PRODUC  \
0                               0                               0   
1                               0                               0   
2                               0                               0   
3                               0                               0   
4                               0                               0   
...                           ...                             ...   
254725                          0                               0   
254726                          0                               0   
254727                          0                               0   
254728                          0                               0   
254729                          0                               0   

        ADULT INCONTINENCE PADS  ...  WRAP  WREATHS  WREATHS/TINSEL/GARLAND  \
0                             0  ...     0        0                       0   
1                             0  ...     0        0                       0   
2                             0  ...     0        0                       0   
3                             0  ...     0        0                       0   
4                             0  ...     0        0                       0   
...                         ...  ...   ...      ...                     ...   
254725                        0  ...     0        0                       0   
254726                        0  ...     0        0                       0   
254727                        0  ...     0        0                       0   
254728                        0  ...     0        0                       0   
254729                        0  ...     0        0                       0   

        WRITING INSTRUMENTS  XMAS PLUSH  YEAST: DRY  YELLOW SUMMER SQUASH  \
0                         0           0           0                     0   
1                         0           0           0                     0   
2                         0           0           0                     0   
3                         0           0           0                     0   
4                         0           0           0                     0   
...                     ...         ...         ...                   ...   
254725                    0           0           0                     0   
254726                    0           0           0                     0   
254727                 

<span style="color:black; background-color:#E6E6FA; padding:2px 4px; border-radius:4px">
<strong> 🥹 실습 때 다루었던 FP-Growth 알고리즘이 참고 자료가 된 이유는 다음 셀 때문이에요... 돌릴 때 조심해주세요... 30분씩 걸릴 때도 있거든요... <strong>
</span>

<span style="color:black; background-color:#fff5b1; padding:2px 4px; border-radius:4px">
<strong> 🤔 "엥? 근데 FP-Growth 알고리즘이 Apriori 알고리즘보다 계산이 빠르다고 하지 않았나?"<strong>
</span>

<span style="color:black; background-color:#E6E6FA; padding:2px 4px; border-radius:4px">
<strong> 🤓 네! 공부를 열심히 하셨군요!? 맞습니다! 이론상 FP-Growth 알고리즘이 Apriori 알고리즘이 계산이 더 빠릅니다!<strong>
</span>

<span style="color:black; background-color:#fff5b1; padding:2px 4px; border-radius:4px">
<strong> 🤔 엥 그럼 왜...? <strong>
</span>

<span style="color:black; background-color:#E6E6FA; padding:2px 4px; border-radius:4px">
<strong> 🤓 사용한 라이브러리의 차이입니다! 저희는 mlxtend 라이브러리를 사용했습니다! <strong>
</span>

<span style="color:black; background-color:#E6E6FA; padding:2px 4px; border-radius:4px">
<strong> 🤓 mlxtend 라이브러리의 경우 Apriori는 Cython으로 최적화 되어 매우 빠르게 작동하지만, FP-Growth는 순수 Python으로 구현하기 때문에 오히려 느릴 수 있습니다. 이 경우 fpgrowth_py 라이브러리를 활용한다면 더 빠르게 작동할 수 있어요~ <strong>
</span>

- FP-Growth 알고리즘으로 빈발 항목 집합 도출 (지지도 0.05% 이상)

In [25]:
from mlxtend.frequent_patterns import fpgrowth

# FP-Growth 알고리즘을 사용해 frequent_itemsets_fp을 생성하세요.
# 조건: 최소지지도: 0.005, use_colnames=True, 입력 데이터는 boolean 타입(astype(bool))
frequent_itemsets_fp = fpgrowth(filtered_onehot.astype(bool), min_support=0.005, use_colnames=True)
frequent_itemsets_fp.head()

support                      itemsets
0  0.118961                     (BANANAS)
1  0.038044  (POTATOES RUSSET (BULK&BAG))
2  0.023715     (ONIONS SWEET (BULK&BAG))
3  0.018816                      (CELERY)
4  0.040133              (HAMBURGER BUNS)

- `association_rules`로 연관 규칙 도출 및 필터링(신뢰도 40% 이상)

In [26]:
# confidence 기준으로 연관 규칙(rules)을 추출하세요.
# 조건: min_threshold=0.4, metric="***"
rules_fp = association_rules(frequent_itemsets_fp, metric="confidence", min_threshold=0.4)

rules_fp.head()

antecedents              consequents  \
0                                          (BANANAS)  (FLUID MILK WHITE ONLY)   
1                       (POTATOES RUSSET (BULK&BAG))  (FLUID MILK WHITE ONLY)   
2              (BANANAS, POTATOES RUSSET (BULK&BAG))  (FLUID MILK WHITE ONLY)   
3  (POTATOES RUSSET (BULK&BAG), MAINSTREAM WHITE ...  (FLUID MILK WHITE ONLY)   
4      (POTATOES RUSSET (BULK&BAG), SHREDDED CHEESE)  (FLUID MILK WHITE ONLY)   

   antecedent support  consequent support   support  confidence      lift  \
0            0.118961            0.240686  0.061402    0.516154  2.144508   
1            0.038044            0.240686  0.019236    0.505624  2.100759   
2            0.010788            0.240686  0.006776    0.628093  2.609593   
3            0.011463            0.240686  0.007577    0.660959  2.746144   
4            0.008519            0.240686  0.005606    0.658065  2.734118   

   representativity  leverage  conviction  zhangs_metric   jaccard  certainty  \
0               1.0  0.032770    1.569328       0.605754  0.205878   0.362785   
1               1.0  0.010079    1.535903       0.544704  0.074129   0.348917   
2               1.0  0.004179    2.041677       0.623525  0.027690   0.510207   
3               1.0  0.004818    2.239592       0.643226  0.030979   0.553490   
4               1.0  0.003556    2.220635       0.639701  0.023013   0.549678   

   kulczynski  
0    0.385633  
1    0.292773  
2    0.328123  
3    0.346219  
4    0.340678

In [27]:
from IPython.display import display

print("🔥 Frequent Itemsets:")
display(frequent_itemsets_fp)

print("\n🔥 Association Rules:")
display(rules_fp)

🔥 Frequent Itemsets:


support                                           itemsets
0     0.118961                                          (BANANAS)
1     0.038044                       (POTATOES RUSSET (BULK&BAG))
2     0.023715                          (ONIONS SWEET (BULK&BAG))
3     0.018816                                           (CELERY)
4     0.040133                                   (HAMBURGER BUNS)
...        ...                                                ...
1475  0.005221             (FLUID MILK WHITE ONLY, PAPER NAPKINS)
1476  0.007267             (STRING CHEESE, FLUID MILK WHITE ONLY)
1477  0.006866   (CANDY BARS (MULTI PACK), FLUID MILK WHITE ONLY)
1478  0.006572  (FLUID MILK WHITE ONLY, ISOTONIC DRINKS SINGLE...
1479  0.005064            (STUFFING MIXES, FLUID MILK WHITE ONLY)

[1480 rows x 2 columns]


🔥 Association Rules:


antecedents  \
0                                            (BANANAS)   
1                         (POTATOES RUSSET (BULK&BAG))   
2                (BANANAS, POTATOES RUSSET (BULK&BAG))   
3    (POTATOES RUSSET (BULK&BAG), MAINSTREAM WHITE ...   
4        (POTATOES RUSSET (BULK&BAG), SHREDDED CHEESE)   
..                                                 ...   
440                     (FACIAL TISSUE & PAPER HANDKE)   
441                                    (PAPER NAPKINS)   
442                                    (STRING CHEESE)   
443                     (ISOTONIC DRINKS SINGLE SERVE)   
444                                   (STUFFING MIXES)   

                 consequents  antecedent support  consequent support  \
0    (FLUID MILK WHITE ONLY)            0.118961            0.240686   
1    (FLUID MILK WHITE ONLY)            0.038044            0.240686   
2    (FLUID MILK WHITE ONLY)            0.010788            0.240686   
3    (FLUID MILK WHITE ONLY)            0.011463            0.240686   
4    (FLUID MILK WHITE ONLY)            0.008519            0.240686   
..                       ...                 ...                 ...   
440  (FLUID MILK WHITE ONLY)            0.018141            0.240686   
441  (FLUID MILK WHITE ONLY)            0.010717            0.240686   
442  (FLUID MILK WHITE ONLY)            0.014133            0.240686   
443  (FLUID MILK WHITE ONLY)            0.015762            0.240686   
444  (FLUID MILK WHITE ONLY)            0.009390            0.240686   

      support  confidence      lift  representativity  leverage  conviction  \
0    0.061402    0.516154  2.144508               1.0  0.032770    1.569328   
1    0.019236    0.505624  2.100759               1.0  0.010079    1.535903   
2    0.006776    0.628093  2.609593               1.0  0.004179    2.041677   
3    0.007577    0.660959  2.746144               1.0  0.004818    2.239592   
4    0.005606    0.658065  2.734118               1.0  0.003556    2.220635   
..        ...         ...       ...               ...       ...         ...   
440  0.008919    0.491668  2.042778               1.0  0.004553    1.493737   
441  0.005221    0.487179  2.024127               1.0  0.002642    1.480662   
442  0.007267    0.514167  2.136253               1.0  0.003865    1.562910   
443  0.006572    0.416936  1.732282               1.0  0.002778    1.302283   
444  0.005064    0.539298  2.240667               1.0  0.002804    1.648166   

     zhangs_metric   jaccard  certainty  kulczynski  
0         0.605754  0.205878   0.362785    0.385633  
1         0.544704  0.074129   0.348917    0.292773  
2         0.623525  0.027690   0.510207    0.328123  
3         0.643226  0.030979   0.553490    0.346219  
4         0.639701  0.023013   0.549678    0.340678  
..             ...       ...        ...         ...  
440       0.519902  0.035690   0.330538    0.264363  
441       0.511441  0.021209   0.324626    0.254436  
442       0.539515  0.029353   0.360168    0.272179  
443       0.429497  0.026300   0.232118    0.222120  
444       0.558953  0.020669   0.393265    0.280169  

[445 rows x 14 columns]

In [28]:
fp_growth = rules_fp.drop(columns=[
    "antecedent support", 
    "consequent support", 
    "representativity", 
    "conviction", 
    "zhangs_metric", 
    "jaccard", 
    "certainty", 
    "kulczynski"
])

- 유의미한 규칙 필터링(향상도 1 이상)

In [29]:
fp_growth = fp_growth[fp_growth['lift'] >= 1]
fp_growth

antecedents  \
0                                            (BANANAS)   
1                         (POTATOES RUSSET (BULK&BAG))   
2                (BANANAS, POTATOES RUSSET (BULK&BAG))   
3    (POTATOES RUSSET (BULK&BAG), MAINSTREAM WHITE ...   
4        (POTATOES RUSSET (BULK&BAG), SHREDDED CHEESE)   
..                                                 ...   
440                     (FACIAL TISSUE & PAPER HANDKE)   
441                                    (PAPER NAPKINS)   
442                                    (STRING CHEESE)   
443                     (ISOTONIC DRINKS SINGLE SERVE)   
444                                   (STUFFING MIXES)   

                 consequents   support  confidence      lift  leverage  
0    (FLUID MILK WHITE ONLY)  0.061402    0.516154  2.144508  0.032770  
1    (FLUID MILK WHITE ONLY)  0.019236    0.505624  2.100759  0.010079  
2    (FLUID MILK WHITE ONLY)  0.006776    0.628093  2.609593  0.004179  
3    (FLUID MILK WHITE ONLY)  0.007577    0.660959  2.746144  0.004818  
4    (FLUID MILK WHITE ONLY)  0.005606    0.658065  2.734118  0.003556  
..                       ...       ...         ...       ...       ...  
440  (FLUID MILK WHITE ONLY)  0.008919    0.491668  2.042778  0.004553  
441  (FLUID MILK WHITE ONLY)  0.005221    0.487179  2.024127  0.002642  
442  (FLUID MILK WHITE ONLY)  0.007267    0.514167  2.136253  0.003865  
443  (FLUID MILK WHITE ONLY)  0.006572    0.416936  1.732282  0.002778  
444  (FLUID MILK WHITE ONLY)  0.005064    0.539298  2.240667  0.002804  

[445 rows x 6 columns]

## **4️⃣ 연관 분석 - 결과 해석**

<span style="color:black; background-color:#E6E6FA; padding:2px 4px; border-radius:4px">
<strong> 🤓 결과를 해석하고 전략을 세우는 게 해당 과제의 핵심이니 꼭!!!! 성의있게 깊게 고민한 흔적을 남겨주세요! <strong>
</span>

### **[연관분석] 지지도 0.9% 이상, 신뢰도 55% 이상, 향상도 1 이상 연관 분석**

In [30]:
results = apriori[(apriori['confidence']>0.55)&(apriori['support']>0.009)&(apriori['lift']>1)]
results

antecedents  \
1                                       (ADULT CEREAL)   
2                                  (ALL FAMILY CEREAL)   
48                                    (CHOCOLATE MILK)   
51                                              (CORN)   
57                    (DAIRY CASE 100% PURE JUICE - O)   
62                                      (EGGS - LARGE)   
64                                    (EGGS - X-LARGE)   
70                               (FRUIT BOWL AND CUPS)   
94                                       (KIDS CEREAL)   
100                           (MACARONI & CHEESE DNRS)   
128                                    (PASTA: CANNED)   
129                                    (PEANUT BUTTER)   
153                    (REFRIGERATED BISCUITS REGULAR)   
186                                 (TOASTER PASTRIES)   
200                    (WAFFLES/PANCAKES/FRENCH TOAST)   
202                               (YOGURT MULTI-PACKS)   
218                       (BANANAS, ALL FAMILY CEREAL)   
235                          (BANANAS, CONDENSED SOUP)   
240          (BANANAS, DAIRY CASE 100% PURE JUICE - O)   
243                            (BANANAS, EGGS - LARGE)   
254                             (KIDS CEREAL, BANANAS)   
257          (BANANAS, MAINSTREAM WHEAT/MULTIGRAIN BR)   
258                  (BANANAS, MAINSTREAM WHITE BREAD)   
269                            (BANANAS, POTATO CHIPS)   
272                                 (BANANAS, PREMIUM)   
277                         (BANANAS, SHREDDED CHEESE)   
280          (BANANAS, SOFT DRINKS 12/18&15PK CAN CAR)   
291                  (BANANAS, YOGURT NOT MULTI-PACKS)   
318  (MAINSTREAM WHITE BREAD, DAIRY CASE 100% PURE ...   
329             (MAINSTREAM WHITE BREAD, EGGS - LARGE)   
349        (MAINSTREAM WHITE BREAD, IWS SINGLE CHEESE)   
356              (KIDS CEREAL, MAINSTREAM WHITE BREAD)   
387             (MAINSTREAM WHITE BREAD, POTATO CHIPS)   
392  (SFT DRNK 2 LITER BTL CARB INCL, MAINSTREAM WH...   
393          (MAINSTREAM WHITE BREAD, SHREDDED CHEESE)   
394  (SNACK CAKE - MULTI PACK, MAINSTREAM WHITE BREAD)   
398  (SOFT DRINKS 12/18&15PK CAN CAR, MAINSTREAM WH...   
428  (SOFT DRINKS 12/18&15PK CAN CAR, SHREDDED CHEESE)   

                 consequents   support  confidence      lift  leverage  
1    (FLUID MILK WHITE ONLY)  0.019154    0.638028  2.650870  0.011928  
2    (FLUID MILK WHITE ONLY)  0.026137    0.649118  2.696946  0.016446  
48   (FLUID MILK WHITE ONLY)  0.021615    0.654308  2.718510  0.013664  
51   (FLUID MILK WHITE ONLY)  0.013379    0.555682  2.308742  0.007584  
57   (FLUID MILK WHITE ONLY)  0.037451    0.588163  2.443691  0.022126  
62   (FLUID MILK WHITE ONLY)  0.029667    0.571547  2.374658  0.017174  
64   (FLUID MILK WHITE ONLY)  0.021752    0.583263  2.423334  0.012776  
70   (FLUID MILK WHITE ONLY)  0.011710    0.572773  2.379748  0.006790  
94   (FLUID MILK WHITE ONLY)  0.033000    0.654316  2.718544  0.020861  
100  (FLUID MILK WHITE ONLY)  0.017100    0.572706  2.379471  0.009914  
128  (FLUID MILK WHITE ONLY)  0.013096    0.569089  2.364444  0.007557  
129  (FLUID MILK WHITE ONLY)  0.014466    0.566749  2.354720  0.008323  
153  (FLUID MILK WHITE ONLY)  0.012331    0.554457  2.303652  0.006978  
186  (FLUID MILK WHITE ONLY)  0.011593    0.575634  2.391635  0.006745  
200  (FLUID MILK WHITE ONLY)  0.011279    0.579000  2.405623  0.006590  
202  (FLUID MILK WHITE ONLY)  0.013418    0.576100  2.393572  0.007812  
218  (FLUID MILK WHITE ONLY)  0.010721    0.716234  2.975800  0.007118  
235  (FLUID MILK WHITE ONLY)  0.009186    0.661017  2.746385  0.005841  
240  (FLUID MILK WHITE ONLY)  0.014451    0.686370  2.851719  0.009383  
243  (FLUID MILK WHITE ONLY)  0.010544    0.698933  2.903918  0.006913  
254  (FLUID MILK WHITE ONLY)  0.009932    0.709877  2.949386  0.006565  
257  (FLUID MILK WHITE ONLY)  0.011314    0.654405  2.718914  0.007153  
258  (FLUID MILK WHITE ONLY)  0.016492    0.663351  2.756081  0.010508  
269  (FLUID MILK WHITE ONLY)

- 위 결과를 지지도, 신뢰도, 향상도 값을 바탕으로 해석해주세요! (두 가지 이상)

<span style="color:black; background-color:#E6E6FA; padding:2px 4px; border-radius:4px">  
<strong> 🤓 [ 예시 ] 'FLUID MILK WHITE ONLY'(우유)는 다양한 품목과 높은 결합 구매 패턴을 보이며, 30개 이상의 제품과 lift 2 이상으로 강한 연관성을 나타낸다. <br> 특히 시리얼 계열인 'ALL FAMILY CEREAL', 'KIDS CEREAL', 'ADULT CEREAL'은 각각 65% 이상의 신뢰도와 2.6~2.7의 lift를 기록해 눈에 띄는 결합 소비가 확인된다. <br> 전반적으로 우유는 시리얼, 과일, 베이커리, 아침 식사류 제품들과 자주 함께 구매되며, 이는 소비자의 식사 준비 맥락과 밀접하게 연결된 구매 경향을 보여준다. <strong>  
</span>

- 해석 1:   
우유는 다양한 품목들과 강한 연관성과 높은 결합 구매 패턴을 보여주고, 그 중 타 상품 + 바나나의 조합 + 우유인 경우를 주목해볼만 하다.  
필터링 된 총 38개의 조합 중, 위의 조합은 12개로 약 30%이상의 비중을 차지하고, 62 ~ 71%의 신뢰도와 2.6 ~ 2.97의 lift를 기록하여 눈에 띄는 결합 소비가 확인된다.  
따라서 우유와 함께 어러 상품을 구매하는 경우 그 조합 속에 바나나가 들어가는 경우가 매우 많고, 이를 통해 @+(바나나+우유)를 같이 섭취하는 소비자들의 소비 패턴을 알 수 있다.  
또한, CEREAL, EGGS, JUICE, BREAD, SOUP, YOGURT등과 많이 소비되는것으로 보아 간단하게 때우는 식사를 위한 소비에 이 조합이 많이 활용되는 것을 알 수 있다.

In [31]:
with_banana=results[results['antecedents'].apply(lambda x: 'BANANAS' in x)]
with_banana.sort_values(by=['confidence','lift'],ascending=False)

antecedents              consequents  \
218               (BANANAS, ALL FAMILY CEREAL)  (FLUID MILK WHITE ONLY)   
254                     (KIDS CEREAL, BANANAS)  (FLUID MILK WHITE ONLY)   
243                    (BANANAS, EGGS - LARGE)  (FLUID MILK WHITE ONLY)   
240  (BANANAS, DAIRY CASE 100% PURE JUICE - O)  (FLUID MILK WHITE ONLY)   
277                 (BANANAS, SHREDDED CHEESE)  (FLUID MILK WHITE ONLY)   
258          (BANANAS, MAINSTREAM WHITE BREAD)  (FLUID MILK WHITE ONLY)   
235                  (BANANAS, CONDENSED SOUP)  (FLUID MILK WHITE ONLY)   
257  (BANANAS, MAINSTREAM WHEAT/MULTIGRAIN BR)  (FLUID MILK WHITE ONLY)   
291          (BANANAS, YOGURT NOT MULTI-PACKS)  (FLUID MILK WHITE ONLY)   
272                         (BANANAS, PREMIUM)  (FLUID MILK WHITE ONLY)   
280  (BANANAS, SOFT DRINKS 12/18&15PK CAN CAR)  (FLUID MILK WHITE ONLY)   
269                    (BANANAS, POTATO CHIPS)  (FLUID MILK WHITE ONLY)   

      support  confidence      lift  leverage  
218  0.010721    0.716234  2.975800  0.007118  
254  0.009932    0.709877  2.949386  0.006565  
243  0.010544    0.698933  2.903918  0.006913  
240  0.014451    0.686370  2.851719  0.009383  
277  0.014023    0.681289  2.830612  0.009069  
258  0.016492    0.663351  2.756081  0.010508  
235  0.009186    0.661017  2.746385  0.005841  
257  0.011314    0.654405  2.718914  0.007153  
291  0.012268    0.643136  2.672095  0.007677  
272  0.009697    0.635452  2.640166  0.006024  
280  0.013159    0.633648  2.632674  0.008161  
269  0.009642    0.625892  2.600448  0.005934

- 해석 2:
우유는 다양한 신선식품들과 강한 연관성과 높은 결합 구매 패턴을 보여주며, 그 중 치즈, 요거트, 주스, 빵 등과 함께 구성된 조합이 특히 눈에 띈다.
필터링된 총 13개의 신선식품 관련 조합은 평균 63.5% 이상의 신뢰도와 2.6~3.0에 달하는 lift를 기록하며, 우유가 냉장 코너 중심 품목들과 함께 자주 구매되는 경향을 확인할 수 있다.
이는 우유가 단일 품목이 아닌 치즈,빵,주스 등과 함께 구성된 아침식사 형태로 소비되는 패턴을 보여주며, 실제 매장 진열 및 프로모션 전략에 있어 강력한 결합 상품군임을 시사한다.

In [35]:
fresh_item=['CHOCOLATE MILK','SOFT DRINKS 12/18&15PK CAN CAR',' SHREDDED CHEESE','IWS SINGLE CHEESE',' YOGURT NOT MULTI-PACKS','DAIRY CASE 100% PURE JUICE - O','YOGURT MULTI-PACKS','REFRIGERATED BISCUITS REGULAR','CHOCOLATE MILK']

fresh_results = results[results['antecedents'].apply(lambda x: any(item.strip() in x for item in fresh_item))]
fresh_results.sort_values(by=['confidence','lift'],ascending=False)

antecedents  \
318  (MAINSTREAM WHITE BREAD, DAIRY CASE 100% PURE ...   
240          (BANANAS, DAIRY CASE 100% PURE JUICE - O)   
277                         (BANANAS, SHREDDED CHEESE)   
393          (MAINSTREAM WHITE BREAD, SHREDDED CHEESE)   
48                                    (CHOCOLATE MILK)   
291                  (BANANAS, YOGURT NOT MULTI-PACKS)   
280          (BANANAS, SOFT DRINKS 12/18&15PK CAN CAR)   
349        (MAINSTREAM WHITE BREAD, IWS SINGLE CHEESE)   
398  (SOFT DRINKS 12/18&15PK CAN CAR, MAINSTREAM WH...   
428  (SOFT DRINKS 12/18&15PK CAN CAR, SHREDDED CHEESE)   
57                    (DAIRY CASE 100% PURE JUICE - O)   
202                               (YOGURT MULTI-PACKS)   
153                    (REFRIGERATED BISCUITS REGULAR)   

                 consequents   support  confidence      lift  leverage  
318  (FLUID MILK WHITE ONLY)  0.010026    0.736235  3.058901  0.006749  
240  (FLUID MILK WHITE ONLY)  0.014451    0.686370  2.851719  0.009383  
277  (FLUID MILK WHITE ONLY)  0.014023    0.681289  2.830612  0.009069  
393  (FLUID MILK WHITE ONLY)  0.013559    0.679520  2.823261  0.008757  
48   (FLUID MILK WHITE ONLY)  0.021615    0.654308  2.718510  0.013664  
291  (FLUID MILK WHITE ONLY)  0.012268    0.643136  2.672095  0.007677  
280  (FLUID MILK WHITE ONLY)  0.013159    0.633648  2.632674  0.008161  
349  (FLUID MILK WHITE ONLY)  0.010466    0.631305  2.622937  0.006476  
398  (FLUID MILK WHITE ONLY)  0.014969    0.621010  2.580163  0.009167  
428  (FLUID MILK WHITE ONLY)  0.010321    0.595740  2.475172  0.006151  
57   (FLUID MILK WHITE ONLY)  0.037451    0.588163  2.443691  0.022126  
202  (FLUID MILK WHITE ONLY)  0.013418    0.576100  2.393572  0.007812  
153  (FLUID MILK WHITE ONLY)  0.012331    0.554457  2.303652  0.006978

## **5️⃣ 연관 분석 - 비즈니스 전략 수립**

- 위 결과해석에 따라 비즈니스 전략을 수립해주세요! (2가지 이상) -> 냅다 GPT만 패서 쓰지 말아주세요 . . .. . 

<span style="color:black; background-color:#E6E6FA; padding:2px 4px; border-radius:4px">  
<strong> 🤓 [ 예시 ] 우유와 함께하는 식탁 큐레이션 존 구성 <strong>  
</span>

**🎯 목적**

* 우유와 자주 함께 구매되는 상품을 모아 **직관적인 구매 유도**
* **객단가 상승**, **편리한 쇼핑 경험 제공**

**🛒 구성 품목**

* **시리얼류**: KIDS / ALL FAMILY / ADULT CEREAL
* **베이커리**: 식빵, 비스킷, 토스터 페이스트리
* **과일류**: 바나나, 컵과일
* **유제품/간편식**: 요거트, 달걀, 마카로니 등

**📍 운영 방법**

* 우유 냉장고 인근에 **"우유와 최고의 궁합!"** 테마존 설치
* POP/QR코드로 **추천 식단**이나 **할인 쿠폰** 제공
* 계절별 테마 구성 (예: 여름=냉과일, 겨울=오트밀)

**📈 기대 효과**

* 우유 결합 구매율 상승
* 연관 제품 매출 증가
* ‘고민 없는 조합’으로 **고객 만족도 향상**



### 전략 1: 바나나&우유

- 목적
    1. 바나나+우유의 동시 진열 및 동선 최적화를 통한 구매 유도
    2. 건강한 간편식 조합 인식 및 연관 상품 매출 확대

- 운영 방법
    1. 동시 진열 및 동선 최적화
        - 바나나와 같은 신성 상품과 우유와 같은 유제품을 같이 공유되는 공간에 진열
        - 이렇게 공유된 공간 -> 계산대로 가는 경로에 빵, 요거트, 달걀, 시리얼과 같은 간편식과 관련된 항목들 진열
    2. 바나나&우유 세트 상품 구성
        - 바나나 한 송이&우유 한 팩을 하나로 패키징
        - 이 패키징을 간편식 코너에도 패널형으로 배치. 간편식 + 바나나&우유 세트를 같이 구매하도록 유도
        - 추가 연관 품목 (ex. 시리얼, 요거트)를 구매하면 간편식 세트로 묶어 할인 혜택 제공

- 기대 효과
    1. 객단가 상승
    2. 소비자의 편의성 증대
    3. 바나나&우유+@ -> 간편식이라는 인식을 소비자에게 심어줌으로써 바나나&우유를 포함하는 간편식이라는 이미지 브랜딩

- 전략 2: 냉장 코너 시너징

- 목적
    1. 우유와 연관도가 높은 신선식품(치즈, 요거트, 주스, 빵 등)과의 결합 소비 유도를 통한 객단가 증가
    2. 냉장 코너 내 효율적인 진열 전략을 통해 소비 동선 내 체류시간 확대 및 충동구매 유발

- 운영 방법
    1. 진열 최적화 및 결합형 진열 전략
        - 우유 주변에 연관성 높은 품목군(요거트, 슬라이스 치즈, 주스, 식빵 등)을 동선에 맞춰 인접 진열
        - 냉장 존 내 우유+빵+요거트를 패키징하여 판매
    2. 세트 상품 및 가격 인센티브 제공
        - “든든한 아침세트 (우유+슬라이스 치즈+토스트용 식빵)”와 같이 세트 상품으로 묶어서 판매
        - 해당 세트에 대해 특정 시간대(이른 오전 or 마감 시간대)에 구매하면 할인 혜택 제공
       

- 기대 효과
    1. 객단가 상승 및 회전율 증가
    2. '우유를 통한 아침 간편식'이라는 이미지 브랜딩

# **🤓 기가 막힌 전략을 제시하는 분께는 행운이 찾아옵니다~🍀**